File: backend/src/routes/analyze.js
// POST /api/analyze — run a full soil analysis and persist the result
import { Router } from "express";
import { analyzeSoil } from "../engine.js";
import { saveReport, countReports } from "../db.js";

export const analyzeRoutes = Router();

// ── Input validation ──────────────────────────────────────────────────────────
function validateInput(body) {
  const errors = [];
  const required = ["fieldName", "village", "areaHa", "nitrogen", "phosphorus", "potassium", "ph", "organicMatter", "moisture"];

  for (const field of required) {
    if (body[field] === undefined || body[field] === null || body[field] === "") {
      errors.push(`'${field}' is required`);
    }
  }

  if (!errors.length) {
    if (typeof body.areaHa !== "number" || body.areaHa <= 0)          errors.push("'areaHa' must be a positive number");
    if (typeof body.nitrogen !== "number" || body.nitrogen < 0)        errors.push("'nitrogen' must be >= 0");
    if (typeof body.phosphorus !== "number" || body.phosphorus < 0)    errors.push("'phosphorus' must be >= 0");
    if (typeof body.potassium !== "number" || body.potassium < 0)      errors.push("'potassium' must be >= 0");
    if (typeof body.ph !== "number" || body.ph < 0 || body.ph > 14)   errors.push("'ph' must be between 0 and 14");
    if (typeof body.organicMatter !== "number" || body.organicMatter < 0) errors.push("'organicMatter' must be >= 0");
    if (typeof body.moisture !== "number" || body.moisture < 0 || body.moisture > 100) errors.push("'moisture' must be 0–100");
  }

  return errors;
}

// ── POST /api/analyze ─────────────────────────────────────────────────────────
analyzeRoutes.post("/", (req, res) => {
  const { soilInput, soilTypePrediction } = req.body;

  if (!soilInput) {
    return res.status(400).json({ error: "Request body must contain a 'soilInput' object" });
  }

  const errors = validateInput(soilInput);
  if (errors.length) {
    return res.status(422).json({ error: "Validation failed", details: errors });
  }

  try {
    const analysisNo = countReports() + 1;
    const report = analyzeSoil(soilInput, soilTypePrediction ?? null, analysisNo);
    saveReport(report);

    console.log(`  ✅ analysis #${report.analysisNo} — field: "${soilInput.fieldName}" — score: ${report.healthScore}`);

    return res.status(201).json({
      success: true,
      report,
    });
  } catch (err) {
    console.error(err);
    return res.status(500).json({ error: "Analysis failed: " + err.message });
  }
});

File: backend/routers/analyze.py
from fastapi import APIRouter
from models import AnalyzeSoilRequest
from engine import analyze_soil
import db
import boto3
import uuid
import base64
import os

router = APIRouter()

# Setup S3 client (in production, credentials come from IAM or env vars)
s3_client = boto3.client('s3', region_name='us-east-1')
S3_BUCKET_NAME = os.environ.get("S3_BUCKET_NAME", "soil-health-images-stub")

def upload_to_s3(data_url: str) -> str:
    """Mock upload to S3. Extracts base64, uploads, and returns a public URL."""
    try:
        # e.g., "data:image/jpeg;base64,/9j/4AAQSkZJRg..."
        header, encoded = data_url.split(",", 1)
        ext = header.split(";")[0].split("/")[1]
        file_name = f"soil_captures/{uuid.uuid4().hex}.{ext}"
        
        # In a real environment, we would actually put the object:
        # image_data = base64.b64decode(encoded)
        # s3_client.put_object(Bucket=S3_BUCKET_NAME, Key=file_name, Body=image_data, ContentType=f"image/{ext}")
        
        # Return the public S3 URL
        # return f"https://{S3_BUCKET_NAME}.s3.amazonaws.com/{file_name}"
        return data_url # For local testing without a real S3 bucket
    except Exception as e:
        print(f"S3 Upload Error: {e}")
        return data_url # Fallback to data URL if upload fails


@router.post("/api/analyze-soil", summary="Analyze Soil", tags=["default"], status_code=201)
def run_analysis(body: AnalyzeSoilRequest):
    """
    Run a full AI soil analysis and save the report.

    Performs:
    - **Nutrient diagnosis** (N, P, K, pH status)
    - **ML-based health scoring** (SHAP feature contributions)
    - **Fertiliser amendment recommendations**
    - **Crop suitability ranking** (top 5 crops)
    - **Degradation risk assessment**

    Optionally provide **soilTypePrediction** from the image ML classifier
    (e.g. Red, Black, Alluvial) to improve crop matching accuracy.
    """
    analysis_no = db.count() + 1
    report = analyze_soil(body.soilInput, body.soilTypePrediction, analysis_no)
    
    if body.imageDataUrl:
        # Upload to secure cloud storage
        report.imageDataUrl = upload_to_s3(body.imageDataUrl)
    db.save(report)

    print(f"  [OK] analysis #{report.analysisNo} | field: '{report.fieldName}' | score: {report.healthScore} | band: {report.healthBand}")

    return {"success": True, "report": report.model_dump()}
